src/parser/core/modules/CastTime.ts

```diff
@@ -90,20 +167,39 @@ export default class CastTime extends Analyser {
 		const defaultCastTime = action?.castTime
 
 		// If there were no modifiers, just use the default (or if the default comes back undefined, shouldn't happen but eh)
-		if (!matchingTimes.length || defaultCastTime == null) {
+		if (!matchingTimes.length || !defaultCastTime) {
 			return defaultCastTime
 		}
 
-		// Find the shortest cast time and assume that.
-		// TODO: Is the above correct? SE probably has some seriously janky ass shit going on their end...
-		return matchingTimes.reduce(
-			(min, ct) => {
-				// if ct.castTime is a negative number, reduce the default by that amount.
-				// Don't let the result drop below 0
-				const castTime = ct.castTime < 0 ? Math.max(0, min + ct.castTime) : ct.castTime
-				return castTime < min ? castTime : min
-			},
-			defaultCastTime,
-		)
+		// Find the largest flat cast time reduction value
+		const flatReduction = matchingTimes.reduce(
+			(reduction, ct) => {
+				if (ct.type === 'percentage' || ct.adjustment > 0) { return reduction }
+				if (ct.adjustment < reduction) { return ct.adjustment }
+				return reduction
+			}, 0)
+		// Find the largest flat cast time increase value
+		const flatIncrease = matchingTimes.reduce(
+			(increase, ct) => {
+				if (ct.type === 'percentage' || ct.adjustment < 0) { return increase }
+				if (ct.adjustment > increase) { return ct.adjustment }
+				return increase
+			}, 0)
+		// Get the total percentage adjustment
+		const percentageAdjustment = matchingTimes.reduce(
+			(adjustment, ct) => {
+				if (ct.type === 'time') { return adjustment }
+				return adjustment * ct.adjustment
+			}, 1)
```

Comment:
this effectively loops the entire array three times, ignoring a different subset each time.
would be cleaner to use a single `for` loop with some conditional on type and adjustment polarity

Reply:
Yeah, that's dumb of me. Fixed.